File: libs/shared/src/client/java.cc
#include "hestia/client/java.h"

#include <stdexcept>

#include "session.h"

namespace hestia::client {
    std::vector<proto::JavaRelease> Java::releases() {
        return session_->call<proto::JavaReleases>().releases;
    }

    std::vector<proto::JavaRuntime> Java::list() {
        return session_->call<proto::JavaList>().runtimes;
    }

    JavaInstallResult Java::install(int major, bool force, const JavaInstallProgressCallback &on_progress) {
        if (major <= 0) {
            const auto releases = session_->call<proto::JavaReleases>().releases;
            if (releases.empty()) {
                throw std::runtime_error("no java releases are available to install");
            }
            major = releases.back().major; // releases arrive sorted ascending
        }
        const auto id = job_id("java");
        const auto done = session_->run_job(
            id, proto::JavaInstallDoneEvent::kTopic, proto::JavaInstallErrorEvent::kTopic,
            [&on_progress](const ipc::Event &event) {
                if (event.topic != proto::JavaInstallProgressEvent::kTopic || !on_progress) return;
                on_progress(event.payload.get<proto::JavaInstallProgressEvent>().progress);
            },
            [&] { session_->call<proto::JavaInstall>({.major = major, .id = id, .force = force}); });
        const auto event = done.get<proto::JavaInstallDoneEvent>();
        return {.runtime = event.runtime, .already_installed = event.already_installed};
    }

    void Java::uninstall(int major) {
        session_->call<proto::JavaUninstall>({.major = major});
    }
} // namespace hestia::client

File: crates/client/src/facades/java.rs
use ipc::errors::IpcError;
use ipc::protocol::Event;
use serde_json::Value;

use crate::session::{job_id, Session};

pub struct Java<'a> {
    pub(crate) session: &'a Session,
}

impl Java<'_> {
    pub async fn releases(&self) -> Result<Vec<proto::java::JavaRelease>, IpcError> {
        Ok(self
            .session
            .call::<proto::java::JavaReleases>(&proto::Empty {})
            .await?
            .releases)
    }

    pub async fn list(&self) -> Result<Vec<proto::java::JavaRuntime>, IpcError> {
        Ok(self
            .session
            .call::<proto::java::JavaList>(&proto::Empty {})
            .await?
            .runtimes)
    }

    pub async fn uninstall(&self, major: i32) -> Result<(), IpcError> {
        let params = proto::java::JavaUninstallParams { major };
        self.session
            .call::<proto::java::JavaUninstall>(&params)
            .await?;
        Ok(())
    }

    /// Install a runtime, blocking until the daemon reports done or error and
    /// forwarding each progress event to `on_progress`. Returns the registered
    /// runtime and whether it was already installed.
    pub async fn install(
        &self,
        major: i32,
        force: bool,
        on_progress: impl Fn(&proto::java::JavaInstallProgress) + Send + Sync + 'static,
    ) -> Result<(proto::java::JavaRuntime, bool), IpcError> {
        use proto::java::{JavaInstall, JavaInstallParams};

        let id = job_id("java-install");
        let on_event = move |event: &Event| {
            if let Ok(progress) =
                serde_json::from_value::<proto::java::JavaInstallProgress>(event.payload.clone())
            {
                on_progress(&progress);
            }
        };

        let session = self.session;
        let install_id = id.clone();
        let payload = self
            .session
            .run_job(
                &id,
                "java.install.done",
                "java.install.error",
                on_event,
                move || async move {
                    let params = JavaInstallParams {
                        major,
                        id: install_id,
                        force,
                    };
                    session.call::<JavaInstall>(&params).await.map(|_| ())
                },
            )
            .await?;

        let runtime =
            serde_json::from_value(payload.get("runtime").cloned().unwrap_or(Value::Null))
                .map_err(|e| IpcError::Malformed(e.to_string()))?;
        let already = payload
            .get("already_installed")
            .and_then(Value::as_bool)
            .unwrap_or(false);
        Ok((runtime, already))
    }
}
